### PR TITLE
Generalize block collection over basis gates

### LINKED ISSUE #4603: Generalize basis gate for Collect2qBlocks and ConsolidateBlocks
let's open an issue for removing these hardcoded gates and generalizing them (and also the ones that are hardcoded in level 3 KAK).

_Originally posted by @ajavadia in https://github.com/Qiskit/qiskit-terra/pull/4446_

### PR DESCRIPTION
**Pull request by kdk**
### Summary

Resolves #4603 .

### Details and comments

- Split listing of optimization passes in level3.py.
- Add euler_basis option to TwoQubitBasisDecomposer.
- Add UnitarySynthesis pass.
- ConsolidateBlocks to always insert UnitaryGate, not decomposition.
- ConsolidateBlocks to determine KAK gate from basis_gates.
- Collect2qBlocks to check gate width instead of using gate name list.

### CODE REVIEW

File: qiskit/transpiler/passes/synthesis/unitary_synthesis.py [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+                if decomposer1q is None:
+                    continue
+                synth_dag = circuit_to_dag(decomposer1q(node.op.to_matrix()))
+            elif len(node.qargs) == 2:
+                if decomposer2q is None:
+                    continue
+                synth_dag = circuit_to_dag(decomposer2q(node.op.to_matrix()))
+            else:
+                synth_dag = circuit_to_dag(
+                    isometry.Isometry(node.op.to_matrix(), 0, 0).definition)
```
Comment (ajavadia) [inline qiskit/transpiler/passes/synthesis/unitary_synthesis.py:112]: Instead of having a simple pass where most of the logic is elsewhere, how about we transfer that logic to the pass? This way we can have different passes (methods) for synthesizing an isometry or statevector or unitary or whatever.

File: test/python/quantum_info/test_synthesis.py [hunk 2]
```diff
@@ -569,6 +572,39 @@ def test_seed_289(self):
...
+            (CXGate(), 'cx'),
+            (CZGate(), 'cz'),
+            (iSwapGate(), 'iswap'),
+            (RXXGate(np.pi/2), 'rxx'),
+        ]
+
+        for basis in product(euler_bases, kak_gates):
+            (euler_basis, oneq_gates), (kak_gate, kak_gate_name) = basis
+
+            with self.subTest(euler_basis=euler_basis, kak_gate=kak_gate):
```
Comment (ajavadia) [inline test/python/quantum_info/test_synthesis.py:597]: should these be ddt?

File: test/python/compiler/test_transpiler.py [hunk 3]
```diff
@@ -719,6 +719,23 @@ def test_block_collection_runs_for_non_cx_bases(self, basis_gates):
...
+    )
+    def test_block_collection_reduces_1q_gate(self, basis_gates, gate_counts):
+        """For synthesis to non-U3 bases, verify we minimize 1q gates."""
+        qc = QuantumCircuit(2)
+        qc.h(0)
+        qc.cx(0, 1)
+
+        out = transpile(qc, basis_gates=basis_gates, optimization_level=3)
+
+        self.assertTrue(Operator(out).equiv(qc))
```
Comment (ajavadia) [inline test/python/compiler/test_transpiler.py:736]:
```suggestion
        self.assertTrue(Operator(out).equiv(Operator(qc)))
```

File: qiskit/transpiler/preset_passmanagers/level3.py [hunk 4]
```diff
@@ -140,7 +140,8 @@ def _swap_condition(property_set):
         _unroll = [Unroller(basis_gates)]
     elif translation_method == 'translator':
         from qiskit.circuit.equivalence_library import SessionEquivalenceLibrary as sel
-        _unroll = [UnrollCustomDefinitions(sel, basis_gates),
+        _unroll = [UnitarySynthesis(basis_gates),
```
Comment (ajavadia) [inline qiskit/transpiler/preset_passmanagers/level3.py:143]: does UnitarySynthesis need to be here given that it is part of `_opt` too? (_unroll passes are called after _opt passes).

File: qiskit/transpiler/passes/optimization/consolidate_blocks.py [hunk 5]
```diff
@@ -126,17 +136,17 @@ def run(self, dag):
...
 
-                    if len(block_qargs) <= 2:
-                        new_dag.apply_operation_back(
-                            self.decomposer(unitary).to_gate(),
-                            sorted(block_qargs, key=lambda x: block_index_map[x]))
-                    else:
-                        new_dag.apply_operation_back(
-                            UnitaryGate(unitary),
-                            sorted(block_qargs, key=lambda x: block_index_map[x]))
+                max_2q_depth = 20  # If depth > 20, there will be 1q gates to consolidate.
```
Comment (ajavadia) [inline qiskit/transpiler/passes/optimization/consolidate_blocks.py:140]: why is this needed?

File: qiskit/transpiler/passes/synthesis/unitary_synthesis.py (hunk 1 above)
Comment (kdk) [reply qiskit/transpiler/passes/synthesis/unitary_synthesis.py:112]: My thinking was to consolidate the synthesis code under `transpiler/synthesis`, and have the passes that run them (in the context of a DAG) under `transpiler/passes/synthesis` since there are cases where users might want to run synthesis outside of a call to `transpile`.

File: test/python/quantum_info/test_synthesis.py (hunk 2 above)
Comment (kdk) [reply test/python/quantum_info/test_synthesis.py:597]: I don't have a preference, I just kept the style of the existing synthesis tests.

File: test/python/compiler/test_transpiler.py (hunk 3 above)
Comment (kdk) [reply test/python/compiler/test_transpiler.py:736]: Is this needed, or for clarity? `equiv` starts with a `Operator(other)` if the input supports it.

File: qiskit/transpiler/preset_passmanagers/level3.py (hunk 4 above)
Comment (kdk) [reply qiskit/transpiler/preset_passmanagers/level3.py:143]: Good catch, this is redundant here.

File: qiskit/transpiler/passes/optimization/consolidate_blocks.py (hunk 5 above)
Comment (kdk) [reply qiskit/transpiler/passes/optimization/consolidate_blocks.py:140]: This is a stopgap to ensure we trigger synthesis in cases where we know there can be a reduction in 1q gate depth (since there's no way to gauge reductions in 1q gate depth without synthesizing the gate). This can be removed with #3658 .

File: test/python/quantum_info/test_synthesis.py (hunk 2 above)
Comment (ajavadia) [reply test/python/quantum_info/test_synthesis.py:597]: I don't mind either way

File: qiskit/transpiler/passes/synthesis/unitary_synthesis.py (hunk 1 above)
Comment (ajavadia) [reply qiskit/transpiler/passes/synthesis/unitary_synthesis.py:112]: ok I'm fine with that. So will you move it under `transpiler/synthesis` here?

File: test/python/compiler/test_transpiler.py (hunk 3 above)
Comment (ajavadia) [reply test/python/compiler/test_transpiler.py:736]: oh I didn't know that function is overloaded.
